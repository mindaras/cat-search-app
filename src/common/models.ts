export interface Option {
  text: string;
  value: string | number;
  selected?: boolean;
}

export type Options = Option[];

export interface Filter {
  options: Option[];
  onChange: (option: Option) => void;
}

export type Filters = Filter[];
