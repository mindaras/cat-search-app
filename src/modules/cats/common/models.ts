export interface Cat {
  id: string;
  url: string;
}

export enum Order {
  Random = "RANDOM",
  Ascending = "ASC",
  Descending = "DESC",
}

export interface Breed {
  id: string;
  name: string;
}
