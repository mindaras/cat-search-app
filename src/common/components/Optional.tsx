interface Props {
  if: boolean;
  children: any;
}

const Optional: React.FC<Props> = (props) => {
  if (props.if) {
    return props.children;
  }

  return null;
};

export { Optional };
