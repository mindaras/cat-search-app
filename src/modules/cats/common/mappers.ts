interface ErrorResponse {
  message: string;
}

function toDataResponse<T>(response: T | ErrorResponse) {
  return ((response as ErrorResponse)?.message ? [] : response) as T;
}

export { toDataResponse };
