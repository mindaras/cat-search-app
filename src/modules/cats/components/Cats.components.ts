import styled from "styled-components";

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: calc(
    100vh - ${(p) => p.theme.sizes.search.height} -
      ${(p) => p.theme.sizes.pagination.height}
  );
`;

export { LoaderContainer };
