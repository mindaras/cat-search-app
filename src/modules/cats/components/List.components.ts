import styled from "styled-components";

const Cat = styled.img`
  width: 25%;
  height: calc(100% / 3);
  object-fit: cover;
  border: 4px solid #fff;
  border-radius: 30px;

  ${(p) => p.theme.media.medium} {
    width: 33.33%;
    height: 30vw;
  }

  ${(p) => p.theme.media.small} {
    width: 100%;
    height: 100vw;
  }
`;

interface ContainerProps {
  single: boolean;
}

const Container = styled.div<ContainerProps>`
  position: relative;
  display: flex;
  flex-wrap: wrap;
  width: 100%;
  height: calc(
    100vh - ${(p) => p.theme.sizes.search.height} -
      ${(p) => p.theme.sizes.pagination.height}
  );

  ${(p) => {
    if (p.single) {
      return `
            justify-content: center;
            align-items: center;

            ${Cat} {
                width: 40vw;
                height: 40vw;

                ${p.theme.media.medium} {
                    width: 50vw;
                    height: 50vw;
                }

                ${p.theme.media.small} {
                    width: 80vw;
                    height: 80vw;
                }
            }
        `;
    }
  }};

  ${(p) => p.theme.media.medium} {
    height: auto;
  }
`;

const NoDataText = styled.div`
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  font-size: 24px;
`;

const NoDataContainer = styled.div`
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: calc(100vh - ${(p) => p.theme.sizes.search.height});

  ${Cat} {
    opacity: 0.1;
  }

  ${(p) => p.theme.media.small} {
    height: auto;
  }
`;

export { Cat, Container, NoDataContainer, NoDataText };
