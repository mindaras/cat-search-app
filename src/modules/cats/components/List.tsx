import { Optional } from "@common/components/Optional";
import * as Types from "../common/models";
import { Cat, Container, NoDataContainer, NoDataText } from "./List.components";
import pawImage from "@common/images/paw.jpeg";

interface Props {
  cats: Types.Cat[];
}

const List: React.FC<Props> = ({ cats }) => (
  <Container single={cats.length < 2}>
    <Optional if={!!cats.length}>
      {cats.map((cat) => (
        <Cat key={cat.id} src={cat.url} alt="Cat" />
      ))}
    </Optional>
    <Optional if={!cats.length}>
      <NoDataContainer>
        <Cat src={pawImage} alt="Paw" />
        <NoDataText>No data</NoDataText>
      </NoDataContainer>
    </Optional>
  </Container>
);

export { List };
