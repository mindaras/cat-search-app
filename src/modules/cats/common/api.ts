import { apiClient } from "@common/network/apiClient";
import { toDataResponse } from "./mappers";
import { Breed, Cat } from "./models";

const baseUrl = "https://api.thecatapi.com/v1";

const options = {
  headers: {
    "x-api-key":
      "live_r8MgPKaoyvCnkZeZk33jzOxd9IzvNK0iu2hA2FQ7eGzLLSgwFsrP81TGuLiKtPfp", // public test key
  },
};

export interface FilterParameters {
  page: number | string;
  order?: string;
  breed?: string;
}

const getAll = async ({ page, order, breed }: FilterParameters) => {
  const limit = 12;
  const url = new URL(baseUrl);
  url.pathname += "/images/search";
  url.searchParams.append("limit", limit.toString());

  if (page) url.searchParams.append("page", page.toString());
  if (order) url.searchParams.append("order", order);
  if (breed) url.searchParams.append("breed_ids", breed);

  try {
    const { data: response, headers } = await apiClient.get<Cat[]>(
      url.toString(),
      options
    );
    const count = headers?.get("pagination-count") || "0";
    const pageCount = Math.floor(parseInt(count, 10) / limit);
    const data = toDataResponse<Cat[]>(response);
    return { data, pageCount };
  } catch (err) {
    return { data: [], pageCount: 0 };
  }
};

interface SearchParameters {
  searchTerm: string;
}

const get = async ({ searchTerm }: SearchParameters) => {
  try {
    const { data } = await apiClient.get<Cat>(
      `${baseUrl}/images/${searchTerm}`
    );

    if (!data?.id) return { data: [] };

    return { data: [data] };
  } catch (err) {
    return { data: [] };
  }
};

const getBreeds = async () => {
  try {
    const { data: response } = await apiClient.get<Breed[]>(
      `${baseUrl}/breeds`
    );
    const data = toDataResponse(response);
    return { data };
  } catch (err) {
    return { data: [] };
  }
};

export const api = { getAll, get, getBreeds };
