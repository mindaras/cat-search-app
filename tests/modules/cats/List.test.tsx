import renderer from "react-test-renderer";
import { List } from "@modules/cats/components/List";
import { Container } from "@testUtils/components/Container";
import { Cat } from "@modules/cats/common/models";

describe("List", () => {
  it("matches a snapshot", () => {
    const cats: Cat[] = [
      { id: "1", url: "/cat1.png" },
      { id: "2", url: "/cat2.png" },
    ];
    const component = renderer.create(
      <Container>
        <List cats={cats} />
      </Container>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
