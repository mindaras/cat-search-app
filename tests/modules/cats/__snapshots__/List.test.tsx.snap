// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List matches a snapshot 1`] = `
<div
  className="sc-gswNZR bprbIW"
>
  <img
    alt="Cat"
    className="sc-bcXHqe dAkOfR"
    src="/cat1.png"
  />
  <img
    alt="Cat"
    className="sc-bcXHqe dAkOfR"
    src="/cat2.png"
  />
</div>
`;
